examples/compute-boids/main.rs

```diff
@@ -0,0 +1,391 @@
+                            },
+                        },
+                    ],
+                }
+            )
+        );
+    }
+
+
+    // run event loop
+    let mut frame_num = 0;
+
+    event_loop.run(move |event, _, control_flow| {
+        *control_flow = if cfg!(feature = "metal-auto-capture") {
+            ControlFlow::Exit
+        } else {
+            ControlFlow::Poll
+        };
+        match event {
+            event::Event::WindowEvent { event, .. } => match event {
+                event::WindowEvent::KeyboardInput {
+                    input:
+                        event::KeyboardInput {
+                            virtual_keycode: Some(event::VirtualKeyCode::Escape),
+                            state: event::ElementState::Pressed,
+                            ..
+                        },
+                    ..
+                }
+                | event::WindowEvent::CloseRequested => {
+                    *control_flow = ControlFlow::Exit;
+                }
+                _ => {}
+            },
+            event::Event::EventsCleared => {
+                let frame = swap_chain.get_next_texture().unwrap();
+                let depth_texture_view = depth_texture.create_default_view();
+
+                // render pass descriptor
+                // TODO dont recreate this every loop (how to borrow swap_chain?)
+                let render_pass_descriptor = wgpu::RenderPassDescriptor {
+                    color_attachments: &[wgpu::RenderPassColorAttachmentDescriptor {
+                        attachment: &frame.view,
+                        resolve_target: None,
+                        load_op: wgpu::LoadOp::Clear,
+                        store_op: wgpu::StoreOp::Store,
+                        clear_color: wgpu::Color::BLACK,
+                    }],
+                    depth_stencil_attachment: Some(wgpu::RenderPassDepthStencilAttachmentDescriptor {
+                        attachment: &depth_texture_view,
+                        depth_load_op: wgpu::LoadOp::Clear, // TODO not sure if Load or Clear
+                        depth_store_op: wgpu::StoreOp::Store,
+                        clear_depth: 1.0,
+                        stencil_load_op: wgpu::LoadOp::Clear, // TODO not sure if Load or Clear
+                        stencil_store_op: wgpu::StoreOp::Store,
+                        clear_stencil: 0,
+                    }),
+                };
+
+                let mut command_encoder =
+                    device.create_command_encoder(&wgpu::CommandEncoderDescriptor { todo: 0 });
+                
+                {
+                    // compute pass
+                    let mut cpass = command_encoder.begin_compute_pass();
+                    cpass.set_pipeline(&compute_pipeline);
+                    cpass.set_bind_group(0, &particle_bind_groups[frame_num % 2], &[]);
+                    cpass.dispatch(NUM_PARTICLES, 1, 1);
```

Comment:
is every particle being processed as a separate compute group? that seems inefficient

Reply:
@kvark I'm somewhat novice to compute programming, I tried an attempt using PARTICLES_PER_INVOCATION and a for loop in the compute shader. I did these in a separate commit. I wasn't able to test the performance differences with this, rather my laptop (osx) would fail with GPU stuck errors when I increased the PARTICLES_PER_INVOCATION number too large.